spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/BindingService.java

```diff
@@ -95,16 +106,56 @@ public BindingService(
 		}
 		validate(consumerProperties);
 		for (String target : bindingTargets) {
-			Binding<T> binding = binder.bindConsumer(target,
-					bindingServiceProperties.getGroup(inputName), input,
-					consumerProperties);
+			Binding<T> binding = doBindConsumer(input, inputName, binder, consumerProperties, target);
 			bindings.add(binding);
 		}
 		bindings = Collections.unmodifiableCollection(bindings);
 		this.consumerBindings.put(inputName, new ArrayList<Binding<?>>(bindings));
 		return bindings;
 	}
 
+	public <T> Binding<T> doBindConsumer(T input, String inputName, Binder<T, ConsumerProperties, ?> binder,
+			ConsumerProperties consumerProperties, String target) {
+		if (this.taskScheduler == null || this.bindingServiceProperties.getBindingRetryInterval() <= 0) {
+			return binder.bindConsumer(target,
+					this.bindingServiceProperties.getGroup(inputName), input,
+					consumerProperties);
+		}
+		else {
+			try {
+				return binder.bindConsumer(target,
+						this.bindingServiceProperties.getGroup(inputName), input,
+						consumerProperties);
+			}
+			catch (RuntimeException e) {
+				LateBinding<T> late = new LateBinding<T>();
+				rescheduleConsumerBinding(input, inputName, binder, consumerProperties, target, late, e);
+				return late;
+			}
+		}
+	}
+
+	public <T> void rescheduleConsumerBinding(final T input, final String inputName,
+			final Binder<T, ConsumerProperties, ?> binder, final ConsumerProperties consumerProperties,
+			final String target, final LateBinding<T> late, RuntimeException exception) {
+		if (exception instanceof IllegalStateException || exception instanceof IllegalArgumentException) {
+			throw exception;
+		}
+		this.log.error("Failed to create consumer binding; retrying in " +
+			this.bindingServiceProperties.getBindingRetryInterval() + " seconds", exception);
+		this.taskScheduler.schedule(() -> {
+			try {
+				late.setDelegate(binder.bindConsumer(target,
+					this.bindingServiceProperties.getGroup(inputName), input,
+					consumerProperties));
```

Comment:
@garyrussell just curious, why are we calling `setDelegate` like this in the case of consumer, but setting it directly in the case of producer.

Reply:
Good catch; we __must__ use the setter to avoid the race.